### PR TITLE
Adds sampler which can sample only parameters needed for site selection

### PR DESCRIPTION
**Pull request by Rosejoycrocker**
- changes to sample() to allow specific model components to be sampled
- addition of sample_site_selection, which allows sampling of only parameters needed for site selection.
- Addresses issue #296

### CODE REVIEW

File: src/io/sampling.jl [hunk 1]
```diff
@@ -65,24 +65,68 @@ end
...
+    return df
+end
+
+"""
+    sample(spec::DataFrame, n::Int, sampler=SobolSample())::DataFrame
+
+Create samples and rescale to distribution defined in the model spec.
+
+# Arguments
+- spec : DataFrame containing model parameter specifications.
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:121]: Add backticks to entries in argument list

File: src/io/sampling.jl [hunk 2]
```diff
@@ -65,24 +65,68 @@ end
...
+function sample(dom::Domain, n::Int, component::Type, sampler=SobolSample())::DataFrame
+    n > 0 ? n : throw(DomainError(n, "`n` must be > 0"))
+
+    spec = component_params(dom.model, component)
+    df = sample(spec, n, sampler)
+    process_inputs!(spec, df)
+
+    # Adjust samples for discrete values using flooring trick
+    # Ensure unguided scenarios do not have superfluous parameter values
+    return df
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:112]: No parameter adjustment or transform to specified distributions? Or is this handled elsewhere?

File: src/io/sampling.jl [hunk 3]
```diff
@@ -65,24 +65,68 @@ end
 
 
 """
-    sample(dom::Domain, n::Int, sampler=SobolSample())::DataFrame
+    sample(dom::Domain, n::Int)::DataFrame
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:68]: Ability to specify a sampler removed?

File: src/io/sampling.jl [hunk 4]
```diff
@@ -131,6 +203,20 @@ function sample_cf(d::Domain, n::Int, sampler=SobolSample())::DataFrame
...
 
+"""
+    adjust_guided_bounds(guided_spec::DataFrame, lower::Int64)::DataFrame
+    
+Adjust lower bound of guided parameter spec to alter sampling range.
+"""
+function adjust_guided_bounds(guided_spec::DataFrame, lower::Int64)::DataFrame
+    guided_col = guided_spec.fieldname .== :guided
+    g_upper = guided_spec[:, :upper_bound][1]
+    guided_spec[guided_col, :val] .= lower
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:214]: If we really really need this as a separate function, you could take the opportunity to optimize - update all values in one go.

```julia
guided_spec[guided_col, [:val, :lower_bound, :bounds, :full_bounds]] .= [lower, lower, (lower, g_upper), (lower, g_upper)]
```

My previous implementation was just trying to be clear about what was happening, but updating values one by one is much slower.

File: src/io/sampling.jl [hunk 5]
```diff
@@ -142,25 +228,18 @@ function sample_guided(d::Domain, n::Int, sampler=SobolSample())::DataFrame
...
     insertcols!(mod_df, :val, :bounds => copy([d.model[:bounds]...]))
-    mod_df[guided_col, :val] .= 1
-    mod_df[guided_col, :lower_bound] .= 1
-    mod_df[guided_col, :bounds] .= [(1, g_upper)]
-    mod_df[guided_col, :full_bounds] .= [(1, g_upper)]
+    mod_df = adjust_guided_bounds(mod_df, 1)
 
     # Sample without unguided, then revert back to original model spec
     ADRIA.update!(d.model, mod_df)
     samples = _sample(d, n, sampler)
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:237]: `_sample()` seems to have been renamed `sample()` at some point, and before alternate distributions were possible.
(meaning this and `sample_cf`, `sample_unguided` are now all broken).

Although we don't seem to do specific cf/unguided/guided sampling any more, might as well take this opportunity and fix it now.

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]

File: src/io/sampling.jl (hunk 5 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl:237]: so everything called sample() should be called _sample()? I was trying to work out where _sample() was defined and could only find sample() so that's what I used...

File: src/io/sampling.jl (hunk 3 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl:68]: Not intentional, this may have been left over from a structural change I tried which I then reverted. I'm make sure this hasn't happened anywhere else

File: src/io/sampling.jl (hunk 4 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl:214]: it doesn't have to be a separate function, it's just it's used 3 time so I thought I'd make a function to avoid replication

File: src/io/sampling.jl (hunk 5 above)
Comment (ConnectedSystems) [reply src/io/sampling.jl:237]: No, it is now `sample()` - we don't need a separate internal function for our purposes hence why it was renamed

File: src/io/sampling.jl (hunk 4 above)
Comment (ConnectedSystems) [reply src/io/sampling.jl:212]: Sure, could you rename the function to indicate it's an internal function (prefix with a `_`) and give it a proper function name? The current one is misleading as it only adjusts the lower bound

File: src/io/sampling.jl (hunk 5 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: right, thanks, I'll change it in sample_cf()

File: src/io/sampling.jl (hunk 4 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: sure

File: src/io/sampling.jl (hunk 2 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl:112]: It's done above this comment, so I'llmove the comment back above

File: src/io/sampling.jl [hunk 6]
```diff
@@ -117,9 +161,37 @@ function sample(dom::Domain, n::Int, sampler=SobolSample(); supported_dists=Dict
...
+    sample_site_selection(d::Domain, n::Int)::DataFrame
+
+Create samples of only site selection parameters and rescale to distribution defined in the model spec.
+
+# Arguments
+- d : Domain.
+- n : number of samples to generate.
+
+"""
+function sample_site_selection(d::Domain, n::Int, sampler=SobolSample())::DataFrame
```
[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:175]: I've changed this so it now constructs a single paramter spec in the same order as the model spec to sample from...by alternate distributions, do you mean the supported_dists variable in sample() ?

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:175]: In a way yes. There's a step to transform the samples to the distributions specified in the model spec that is missing.

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:175]: See from L94 to L99 in the below

Triangular distributions are handled as a special case because a three-element entry, $a, b, c$ is expected where these relate to the lower and upper bounds and the location of the peak in percentile terms.

https://github.com/open-AIMS/ADRIA.jl/blob/7b50c835411618f8698493cb8cd62f6a00be20e1/src/io/sampling.jl#L94

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:175]: I can see this section of code still on L136 to L142 
https://github.com/open-AIMS/ADRIA.jl/blob/6b08bde0b652ef58073fec640211d83c0d4f1e0a/src/io/sampling.jl#L136

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:175]: Okay, I see the flow now. Thank you.

File: test/site_selection.jl [hunk 7]
```diff
@@ -34,12 +34,13 @@ end
...
     dom = ADRIA.load_domain(joinpath(@__DIR__, "..", "examples", "Example_domain"), 45)
-    criteria_df = ADRIA.sample_site_selection(dom, 5)  # get scenario dataframe
+    N = 2^3
+    criteria_df = ADRIA.sample_site_selection(dom, N)  # get scenario dataframe
 
     area_to_seed = 1.5 * 10^-6  # area of seeded corals in km^2
     ts = 5  # time step to perform site selection at
 
-    sum_cover = 0.1 .* ones(5, size(dom.site_data, 1))
+    sum_cover = 0.1 .* ones(N, size(dom.site_data, 1))
```
Comment (ConnectedSystems) [inline test/site_selection.jl:43]: Didn't we replace this line with `fill(0.1, N, n_locations(dom))` ?

Comment (Rosejoycrocker) [reply test/site_selection.jl:43]: I think we might have missed this (I did rebase with Main, so if it was replace it shoiuld be in here), I'll change it.

**Comment by Rosejoycrocker** [comment]
I think I've addressed your comments, let me know if I've missed anything

**Comment by Rosejoycrocker** [comment]
Actually, still need to fix something with the datatype of the process_inputs! function... is there a data type that covers vectors and tuples? Like AbstractVector, or something?

File: src/io/sampling.jl [hunk 8]
```diff
@@ -117,20 +159,58 @@ function sample(dom::Domain, n::Int, sampler=SobolSample(); supported_dists=Dict
...
 
+"""
+    adjust_guided_bounds(guided_spec::DataFrame, lower::Int64)::DataFrame
+    
+Adjust lower bound of guided parameter spec to alter sampling range.
+"""
+function _adjust_guided_lower_bound(spec_df::DataFrame, lower::Int64)::DataFrame
+    guided_col = spec_df.fieldname .== :guided
+    g_upper = spec_df[:, :upper_bound][1]
+    spec_df[guided_col, [:val, :lower_bound, :bounds, :full_bounds]] .= reshape([lower, lower, (lower, g_upper), (lower, g_upper)], (1, 4))
```
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:210]: Streamline this please:

```julia
g_upper = spec_df[:, :upper_bound][1]

# Should be:
spec_df[guided_col, :upper_bound]
```

File: src/io/sampling.jl [hunk 9]
```diff
@@ -117,20 +159,58 @@ function sample(dom::Domain, n::Int, sampler=SobolSample(); supported_dists=Dict
...
+    process_inputs!(sample_df, site_selection_sample)
+    return site_selection_sample
+end
 
 """
     sample_cf(d::Domain, n::Int, sampler=SobolSample())::DataFrame
     
 Generate only counterfactual scenarios using any sampler from QuasiMonteCarlo.jl
 """
 function sample_cf(d::Domain, n::Int, sampler=SobolSample())::DataFrame
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:197]: I'll add tests for `sample_cf` and `sample_guided`

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Missed some earlier requests for changes and found some new ones.

I will add tests for the alternate scenario sampling functions.

**Comment by ConnectedSystems** [comment]
> Actually, still need to fix something with the datatype of the process_inputs! function... is there a data type that covers vectors and tuples? Like AbstractVector, or something?

Just convert to a tuple?

File: src/io/sampling.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:210]: Also just realised this modifies its input (`spec_df`) so please add `!` to the function name

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:212]: It was throwing a size mismatch error because LHS was (1,4) and RHS was (4,)

File: src/io/sampling.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:177]: which parts are not resolved sorry?

File: src/io/sampling.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:211]: Yes - so column data (vector) was being provided to update a row.

No need to complicate things with an additional call, just provide the row:

```julia
julia> df = DataFrame(rand(3,5), :auto)
3×5 DataFrame
 Row │ x1          x2        x3        x4           x5       
     │ Float64     Float64   Float64   Float64      Float64  
─────┼───────────────────────────────────────────────────────
   1 │ 0.00723953  0.202598  0.318617  0.000524996  0.554819
   2 │ 0.687003    0.640692  0.812998  0.351252     0.960522
   3 │ 0.393727    0.37577   0.435917  0.119245     0.825794

julia> df[3, :] .= [1; 2; 3; 4; 5]
DataFrameRow
 Row │ x1       x2       x3       x4       x5      
     │ Float64  Float64  Float64  Float64  Float64 
─────┼─────────────────────────────────────────────
   3 │     1.0      2.0      3.0      4.0      5.0

# OR
julia> df[3, :] .= [1 2 3 4 5]'
DataFrameRow
 Row │ x1       x2       x3       x4       x5      
     │ Float64  Float64  Float64  Float64  Float64 
─────┼─────────────────────────────────────────────
   3 │     1.0      2.0      3.0      4.0      5.0

julia> df
3×5 DataFrame
 Row │ x1          x2        x3        x4           x5       
     │ Float64     Float64   Float64   Float64      Float64  
─────┼───────────────────────────────────────────────────────
   1 │ 0.00723953  0.202598  0.318617  0.000524996  0.554819
   2 │ 0.687003    0.640692  0.812998  0.351252     0.960522
   3 │ 1.0         2.0       3.0       4.0          5.0
```

File: src/io/sampling.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:177]: > The approach of combining separate samples breaks the Sobol' scheme (Sobol' sampling is not pseudo-random sampling).

> Collate all parameters for the desired combination of components first, then sample.

> I would extend the sample() function to accept one or more components for this.

```julia
function sample(d, n, components...; sampler=SobolSample())
    # Extract component parameters and process in one go in here
end
```

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:177]: sample_site_selection no longer combines sample though, it first collects the components into a spec dataframe and then samples everything at once. Do you mean just the more than one components bit for sample? It currently does one component but I can add more

File: src/io/sampling.jl [hunk 10]
```diff
@@ -142,25 +223,17 @@ function sample_guided(d::Domain, n::Int, sampler=SobolSample())::DataFrame
...
-    samples = _sample(d, n, sampler)
-    samples = adjust_samples(d, samples)
+    samples = adjust_samples(d, sample(d, n, sampler))
 
     # Note: updating with spec_df does not work.
-    mod_df[guided_col, :val] .= 0
-    mod_df[guided_col, :lower_bound] .= 0
-    mod_df[guided_col, :bounds] .= [(0, g_upper)]
-    mod_df[guided_col, :full_bounds] .= [(0, g_upper)]
+    mod_df = _adjust_guided_lower_bound(mod_df, 0)
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:235]: Found a bug: this should be -1 (reverting back to the original value) but let me fix it on my end.

File: src/io/sampling.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:177]: To be more specific, instead of this, where individual components are being extracted one by one then concatenated together:

```julia
crit_spec = component_params(d.model, Criteria)
env_spec = component_params(d.model, EnvironmentalLayer)

int_spec = component_params(d.model, Intervention)

# ...

sample_df = vcat(vcat(env_spec, guided_spec), crit_spec)
```

Make it so you can do all that in one go (and avoid that awkward `vcat(vcat())`

```julia
sub_spec = component_params(d.model, EnvironmentalLayer, Intervention, Criteria)
site_selection_sample = sample(sub_spec, n, sampler)
# ...
```

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:177]: Past Takuya already thought of this use case, so please make use of this method:

https://github.com/open-AIMS/ADRIA.jl/blob/043f23e825cb96a2a5ae4a92bc6573a2c0862aa0/src/Domain.jl#L390

File: src/io/sampling.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:212]: Yeah I tried the semicolons but it didn't change the size (maybe cause of the Tuples insie?) 
size_output

The transpose seems to change the size but then when I try the allocation it throws an error, something about how the adjoint doesn't exist for vectors of tuples...

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:210]: As above, use semi-colons instead of commas:

`df[3, :] .= [1; 2; 3; 4; 5]`

Or remove the commas and transpose:

`df[3, :] .= [1 2 3 4 5]'`

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:210]: As in the screenshot, the semicolons don't seem to change the size and the transpose errors
transpose_error

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:210]: can you see the screen shots?

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:210]: The screen shots show you're using commas?

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:210]: in line 3 I use semicolons

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:210]: see these 2 lines give the same size
size_output2

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:210]: image

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:210]: dim_error_semicols

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl:210]: image

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl:210]: oh so you don't need commas at all! My comma version was erroring as well

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl]: image

Have you kept your environment up to date?

[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/io/sampling.jl]: Also, can you check `guided_col`? It should select a single row.

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: guided_col is a one-element BitVector

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: it seems like it was g_upper that was causing the error, it was a vector of one element for some reason

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: updating everything now just incase

[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: works now by making `g_upper = spec_df[guided_col, :upper_bound][1]`

**Review by ConnectedSystems — APPROVED** [review]

**Comment by ConnectedSystems** [comment]
Thanks for doing the hard work on this. I've made some adjustments to simplify things - particularly the flow between each sampling use case and how parameter bounds are adjusted.

I'd like you to have a look over the code when able to get a sense of what the changes were.
The very last thing I'll ask for before I merge is that you look at the tests I've written - perhaps there are other edge cases you'd like to check.

Note that I've associated #296 to be closed with this PR manually. As mentioned previously, GitHub doesn't seem to recognize the closing keywords if it is not in the main text (i.e., it has to be in the main body proper, it can't be in a list or the title).

FYI - I think I mentioned this a while back - while I don't mind, it is a little unusual to use list format in the PR description. Typically you would describe _how_ the PR resolves a given issue, rather than an overview of changes (which we'd able to discern from your commit logs anyway). You can look at other projects on GitHub to see what the norm is.

**Comment by Rosejoycrocker** [comment]
Thanks for cleaning it up :) I'm looking over tests and things now

**Comment by Rosejoycrocker** [comment]
(also sorry about the dotpoints, I'll remember this next time)

**Comment by Rosejoycrocker** [comment]
Looks good to me! I can't think of any other cases we'd need to catch.
